### PR TITLE
[3546][ADD] stock_lot_suffix

### PR DESCRIPTION
**Pull request by kanda999**
[3546](https://www.quartile.co/web?#id=3546&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

File: stock_lot_suffix/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Stock Lot Suffix",
+    "version": "16.0.1.0.0",
+    "author": "Quartile Limited",
+    "website": "https://www.quartile.co",
+    "category": "",
+    "license": "AGPL-3",
+    "depends": ["purchase", "stock", "analytic_mixin_analytic_account"],
```
Comment (yostashiro) [inline stock_lot_suffix/__manifest__.py]:
```suggestion
    "depends": ["purchase.stock", "analytic_mixin_analytic_account"],
```

File: stock_lot_suffix/models/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _set_channel_category(self):
+        move_lines = self.env["stock.move.line"].search([("picking_id", "=", self.id)])
+        lot_ids = move_lines.mapped("lot_id").ids
+        if lot_ids:
+            lots = self.env["stock.lot"].browse(lot_ids)
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_picking.py]: `self` here is a recordset, therefore you need to loop over it.

```suggestion
        for pick in self:
            lots = self.move_line_ids.lot_id
            if lots:
```

File: stock_lot_suffix/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+This module can add a suffix to the stock_lot. The display of the name in the stock_lot changes by this module when a specific partner or analytic plan is used.
```
Comment (yostashiro) [inline stock_lot_suffix/readme/DESCRIPTION.rst]: Please use the `stock.lot` nomenclature (instead of `stock_lot`) for model technical names.

**Review by yostashiro — CHANGES_REQUESTED** [review]
Partial review.

File: stock_lot_suffix/models/account_analytic_account.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    is_suffix = fields.Boolean("Suffix")
```
Comment (yostashiro) [inline stock_lot_suffix/models/account_analytic_account.py]: Didn't we talk about making it a char field?

File: stock_lot_suffix/models/account_analytic_account.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    analytic_category = fields.Char()
```
Comment (yostashiro) [inline stock_lot_suffix/models/account_analytic_account.py]: The name is obscure.  I think we should name it as `lot_suffix` or something IMO.

File: stock_lot_suffix/models/stock_lot.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    def name_get(self):
+        name_list = super(StockLot, self).name_get()
+        new_name_list = []
+        for name in name_list:
+            rec = self.browse(name[0])
+            channel = rec.channel_category
+            analytic = rec.analytic_category
+            new_name = name[1]
+            if analytic:
+                new_name += analytic
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_lot.py]:
```suggestion
            new_name = name[1]
            if rec.analytic_category:
                new_name += rec.analytic_category
```
A minor point.  This is a bit more readable with less LOC.

File: stock_lot_suffix/models/stock_picking.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+from odoo import models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _set_channel_category(self):
+        for picking in self:
+            lots = self.move_line_ids.lot_id
+            if lots:
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_picking.py]: Reduce the `if` ladder.

```suggestion
            if not lots:
                continue
```

File: stock_lot_suffix/models/stock_picking.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+                    for ad in picking.purchase_id.analytic_account_ids:
+                        if ad.analytic_category:
+                            names.append(ad.analytic_category)
+                    if names:
+                        update_values["analytic_category"] = "-".join(names)
+                lots.write(update_values)
+
+    def _action_done(self):
+        self._set_channel_category()
+        return super()._action_done()
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_picking.py]: Can you actually consider doing this at stock.move.line level, which I think should be more simple and flexible.

File: stock_lot_suffix/models/stock_move_line.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            for ad in self.move_id.picking_id.purchase_id.analytic_account_ids:
+                if ad.lot_suffix:
+                    names.append(ad.lot_suffix)
+            if names:
+                update_values["lot_suffix"] = "-".join(names)
+        lots.write(update_values)
+
+    def _action_done(self):
+        self._set_lot_suffix()
+        return super()._action_done()
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_move_line.py]:
```suggestion
    def _action_done(self):
        res = super()._action_done()
        for ml in self:
            if not ml.lot_id:
                continue
            vals = {}
            purchase_line = ml.move_id.purchase_line_id
            vals["channel_category"] = purchase_line.order_id.channel_category
            # We assume that there is only one analytic account with lot_suffix if any.
            vals["lot_suffix"] = purchase_line.analytic_account_ids.mapped("lot_suffix")[:1]
            ml.lot_id.write(vals)
        return res
```

**Comment by kanda999** [comment]
This will cause the test to fail, but I can do gitaggregate.

File: stock_lot_suffix/models/stock_move_line.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+            vals = {}
+            purchase_line = ml.move_id.purchase_line_id
+            vals["channel_category"] = purchase_line.order_id.channel_category
+            # We assume that there is only one analytic account with lot_suffix if any.
+            lot_suffixes = [
+                suffix
+                for suffix in purchase_line.analytic_account_ids.mapped("lot_suffix")
+                if suffix
+            ]
+            vals["lot_suffix"] = lot_suffixes[0] if lot_suffixes else None
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_move_line.py]:
```suggestion
            analytic_account = purchase_line.analytic_account_ids.filtered(lambda x: x.lot_suffix)
            if analytic_account:
                vals["lot_suffix"] = analytic_account[0].lot_suffix
```

File: stock_lot_suffix/models/stock_lot.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class StockLot(models.Model):
+    _inherit = "stock.lot"
+
+    channel_category = fields.Char()
+    lot_suffix = fields.Char()
+
+    def name_get(self):
+        name_list = super(StockLot, self).name_get()
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_lot.py]:
```suggestion
        name_list = super().name_get()
```

File: stock_lot_suffix/models/stock_lot.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        new_name_list = []
+        for name in name_list:
+            rec = self.browse(name[0])
+            new_name = name[1]
+            if rec.lot_suffix:
+                new_name += rec.lot_suffix
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_lot.py]:
```suggestion
                new_name += "-" + rec.lot_suffix
```

File: stock_lot_suffix/models/stock_lot.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+            if rec.channel_category:
+                new_name += rec.channel_category
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_lot.py]:
```suggestion
                new_name += "-" + rec.channel_category
```

File: stock_lot_suffix/readme/DESCRIPTION.rst [hunk 14]
```diff
@@ -0,0 +1 @@
+This module can add a suffix to the stock.lot. The display of the name in the stock.lot changes by this module when a specific partner or analytic plan is used.
```
Comment (yostashiro) [inline stock_lot_suffix/readme/DESCRIPTION.rst]: The explanation doesn't sound accurate.  Please improve this.

File: stock_lot_suffix/views/res_partner_views.xml
Comment (yostashiro) [inline stock_lot_suffix/views/res_partner_views.xml:1]: No need to show the field in the list view?

File: stock_lot_suffix/models/purchase_order.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+    _name = "purchase.order"
+    _inherit = ["purchase.order", "analytic.mixin"]
+
+    channel_category = fields.Char()
+
+    @api.onchange("partner_id")
+    def _onchange_channel_category(self):
+        channel = self.partner_id.channel_category
+        if channel:
+            self.channel_category = channel
```
Comment (yostashiro) [inline stock_lot_suffix/models/purchase_order.py]:
```suggestion
    def _onchange_partner_id(self):
        if self.partner_id:
            self.channel_category = self.partner_id.channel_category
```

File: stock_lot_suffix/__manifest__.py [hunk 16]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Stock Lot Suffix",
+    "version": "16.0.1.0.0",
+    "author": "Quartile Limited",
+    "website": "https://www.quartile.co",
+    "category": "",
```
Comment (yostashiro) [inline stock_lot_suffix/__manifest__.py]:
```suggestion
    "category": "Inventory",
```

File: stock_lot_suffix/models/stock_move_line.py [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+                lambda x: x.lot_suffix
+            )
+            if analytic_account:
+                vals["lot_suffix"] = analytic_account[0].lot_suffix
+            # lot_suffixes = [
+            #     suffix
+            #     for suffix in purchase_line.analytic_account_ids.mapped("lot_suffix")
+            #     if suffix
+            # ]
+            # vals["lot_suffix"] = lot_suffixes[0] if lot_suffixes else None
```
Comment (yostashiro) [inline stock_lot_suffix/models/stock_move_line.py]:
```suggestion
```

File: stock_lot_suffix/readme/DESCRIPTION.rst [hunk 18]
```diff
@@ -0,0 +1,6 @@
+This module provides the functionality to manage lot references by adding suffixes.
```
Comment (yostashiro) [inline stock_lot_suffix/readme/DESCRIPTION.rst]:
```suggestion
This module provides the functionality to add suffixes to the display name of lots/serials.
```

File: stock_lot_suffix/readme/DESCRIPTION.rst [hunk 19]
```diff
@@ -0,0 +1,6 @@
...
+
+Suffixes are added in the following scenarios:
+
+- When a purchase order is created with a supplier that the channel_category is set.
+- When a purchase order is created using an analytic plan that the lot suffix is set.
```
Comment (yostashiro) [inline stock_lot_suffix/readme/DESCRIPTION.rst]:
```suggestion
Suffixes are added to lots/serials created from purchase receipts with following conditions:

- Purchase order has the channel_category set.
- Purchase order line has an analytic account with lot suffix.
```

**Review by yostashiro — APPROVED** [review]
Code review.

**Pull request merged**